### ISSUE TITLE
[Feature Request] Unstructured Interpolation for PyTorch Tensor

### ISSUE THREAD

**Issue by lijunzh** [pytorch/pytorch#1552]
I recently discovered the PyTorch Tensor and am very excited about its GPU support. However, it seems to support no interpolation algorithms as MATLAB / Numpy does. My application requires a pre-processing step using linear interpolation of the input data. It is OK now that this pre-processing is done in CPU using _scipy.interpolate_, but it will be nicer if there is something inside PyTorch Tensor that supports doing that operation inside GPU since they will be load into GPU eventually. 

Since I am pretty new to PyTorch, I may be wrong that interpolation is an existing support operator (I did check the doc and found nothing about it). If not, do you guys think it is a good operation to be done in GPU? Is it something that you guys would like to  **include in PyTorch**. If so, I can open a PR and start to working on it.

cc @ezyang @gchanan @zou3519 @mruberry @rgommers @heitorschueroff

**Comment by kvrd18** [comment]
I believe you can work with [UpSampling](http://pytorch.org/docs/nn#torch.nn.UpsamplingBilinear2d) for GPU accelerated interpolation of your images.

**Comment by lijunzh** [comment]
@kvrd18 Thanks a lot for answering my question. I didn't made it clear in my original post. I actually only need to repeatedly shift the Tensor to some off-grid point. For example, given 1-D signal sampled at integer point (0, 1, 2, etc.), we want the interpolated values at non-integer points (0.2, 1.4, 2.9, etc). This is an interpolation problem in general. 

[UpSampling](http://pytorch.org/docs/nn#torch.nn.UpsamplingBilinear2d) should solve the problem with maybe an unnecessary cost. One can easily upsample the tensor high enough and pick the sample point as needed. However, it will be nice if the those **Interpolators** (nearest, linear, bilinear, cubic spline etc.) are available directly in the PyTorch API just like what's offered in [Numpy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.interp.html)/[Scipy](https://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html) and [MATLAB](https://www.mathworks.com/help/matlab/interpolation.html;jsessionid=2e3857153578f1c032555a80441c). Maybe it does not make too much sense for neural networks to have this kind of functions, but it will help when you use PyTorch as a general numpy GPU support. 

I do have some ideas about implementing this and would like to ask if people in PyTorch community if it is a good idea to include those features. @apaszke and @soumith  Forgive me to tag you guys directly if you are not interested in this, but I wanna hear your opinions. Do you guys think PyTorch should support more Numpy features and make it more of a general purpose GPU computing module or it mostly intend to be a neural network tool that complete with [TensorFlow](https://www.tensorflow.org) with dynamic computational graph.  Thanks a lot.

**Comment by lijunzh** [comment]
It seems that there is not much interest in this feature in PyTorch community. I will start a new project instead to enable it outside PyCharm.

**Comment by soumith** [comment]
in the long term i'd def like to see interpolation algorithms in pytorch. I'll keep this issue open.

**Comment by davidbau** [comment]
+1.  (Just spent a couple days working around the lack of interpolation in pytorch.)  There are several forms of interpolation that might be implemented, and they are all useful in different situations.  Here is a list of what's in numpy/scipy. The first three have been useful to me; it would be super-nice to have GPU versions of them.

numpy.interp - 1d interpolation - nice for representing custom functions (e.g., probability densities) built from data.
scipy.ndimage.zoom - similar to torch.nn.UpSampling, except it supports non-integer zoom ratio.  Handy for the common case of scaling an image.
scipy.interpolate.RectBivariateSpline/RegularGridInterpolator - allows generalized and irregular grid points, instead of assuming uniform edge-to-edge zooming.  Useful when being careful but while the data is still at grid points.
scipy.interpolate.griddata etc - totally generic N-D interpolation by tessellating the space with simplexes.  I haven't had an occasion to need this.

**Comment by soumith** [comment]
we have a 2D grid interpolator now, via: http://pytorch.org/docs/0.3.0/nn.html#torch.nn.functional.grid_sample

it works for 4D images (NxCxHxW) and works on HxW dimensions

**Comment by JunjieHu** [comment]
@soumith   Thanks for making the grid_sample function! 

I am confused of the usage of grid in this function. Could you give me an example on how to use this function?

I have a torch.autograd.Variable with the size (batch_size, H_in, W_in). I want to resize this Variable and get the new Variable with the size (batch_size, H_out, W_out). Do you know how to do that?

PS: I found torchvision.transforms has the Resize() function, but it only works for PIL images or numpy array. I need a similar resize function which can be applied directly on the autograd.Variable. Could you confirm that the grid_sample function can work for my situation? Thanks in advance!!

**Comment by peteflorence** [comment]
Hi all -- for a project I'm working on, I made a simple PyTorch bilinear interpolation function, benchmarked it vs. a comparable numpy implementation, and also wrapped the `nn.functional.grid_sample()` function to support my same interface.

In case it's useful to anybody in the PyTorch community I've documented it here: https://gist.github.com/peteflorence/a1da2c759ca1ac2b74af9a83f69ce20e

@JunjieHu this includes an example of how to use `nn.functional.grid_sample()` (see the last section)

**Comment by ailzhang** [comment]
We now have nn.functional.interpolate that does interpolations except for bicubic method. https://pytorch.org/docs/master/nn.html?highlight=interpolate#torch.nn.functional.interpolate
We have another issue for bicubic open here. #918
Closing this for now. Feel free to reopen if you have any questions

**Comment by aliutkus** [comment]
I feel like this is not what I would call a general interpolation method, because it assumes some constant up-downscaling factor.

In the case of 1d, I think that there is a need for a function that would do interpolate(xd,yd,x) where xd are sorted x-values, yd are the corresponding output values, and x are the points you want the linear interpolation from.

This is similar to the numpy interp1d function.
This doesn't seem to be existing in pytorch. I could code it myself, but I find another problem, which is equivalently:
* there is no equivalent to numpy digitize function
* histc doesn't take custom bins boundaries

Right now, it looks like these interpolation functions are all kind of applying deformations to the input. The sample_grid function doesn't look like what I'm looking at neither. 

So I'm not really sure that these developments actually addressed the issue of the lack of general purpose interpolation functions in pytorch

thanks a lot

**Comment by soumith** [comment]
@aliutkus you make a good point. I'm reopening this for us to increase the scope of the feature and get the implementations.

**Comment by aliutkus** [comment]
OK, so, I have spend some time these last days implementing a pytorch extension for doing 1-D interpolation.

To this purpose:
1. I implemented a [pytorch-searchsorted](https://github.com/aliutkus/pytorch-searchsorted) function, that looks for desired values in sorted arrays. This may be useful for other tasks than just interpolation
2. I implemented a [pytorch-interp1d](https://github.com/aliutkus/pytorch-interp1d) function, that depends on 1. and that achieves parallelized GPU 1-D interpolation.

Depending on the desired dimensionality of the problem, the speedup gained by using the GPU can be quite cool:
```
Solving 100000 interpolation problems: each with 100 observations and 30 desired values
CPU: 8060.260ms, GPU: 70.735ms, error: 0.000000%.
```

I hope this helps some one !

Cheers

**Comment by Harper111** [comment]
Hi all. Just like aliutkus mentioned, current interpolation in pytorch always assumed regular sampling grid. 

Besides 1D case, I think it is also necessary to extend the unstructured interpolation to 2D. Because sometimes we need forward warping, not just backward warping.

I want to ask if this feature has been achieved in pytorch or if it would be possible in future?

**Comment by Cuiyirui** [comment]
Well actually, you can also use np.interp() to process torch type data. But the type of return value is numpy. Then you change it to torch type~

**Comment by mileslucas** [comment]
Hello, I am also interested in interpolation schemes.

In my current work there is a need for 1D interpolation and we currently use scipy's `InterpolatedUnivariateSpline` interpolator. It seems like the `grid_sample` method is not quite what we need.

**Comment by Haydnspass** [comment]
Are there updates on this or any way I could help?
I would need to compute cubic spline coefficients on 'calibration' data and then evaluate the function on arbitrary points xq. 
So it would be nice to construct the interpolation and then query:

```
interp_model = interpN(x, y, mode='cubic-spline')
yq = interp_model(xq)
```
where x is N x D-dimensional and y N x 1 dimensional

**Comment by aliutkus** [comment]
I updated my code for the interp1d code to be compatible with pytorch 1.0, it is there:

https://github.com/aliutkus/torchinterp1d

**Comment by zhangguanheng66** [comment]
I'm going to take a look and see how we can implement this in pytorch core.

**Comment by lijunzh** [comment]
@zhangguanheng66 That will be very helpful. Now, I load my data into CPU and do the interpolation by NumPy and then load it back to GPU. I will save much time for me to do everything inside the GPU.

**Comment by ailzhang** [comment]
@zhangguanheng66 IIRC there're ongoing work migrating TH implementations to Aten for many interpolation modes. Good to check with @ezyang and @fmassa on this. :D

**Comment by zhangguanheng66** [comment]
@ailzhang thanks for the remind. Will talk to Ed next week.

**Comment by mechamoth** [comment]
Just as a heads up, some of the fast.ai students ended up replicating some of [TensorFlow's image warping ops](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/image/python/ops) (along with bilinear interpolation) [in a standalone repo](https://github.com/zcaceres/spec_augment/blob/master/SparseImageWarp.ipynb). It's a pretty direct port and probably doesn't need to be as complex as it currently is, but it's something.

**Comment by ezyang** [comment]
For the record, it's this PR: #19630. Should be landing soon.

**Comment by fmassa** [comment]
The difference between the interpolate from this issue is that we don't assume that the interpolation grid is uniform.
So this is not exactly equivalent to the aforementioned PRs.

One way of seeing this is as follows: this issue asks for a more generic way of performing 1d interpolation, where instead of assuming an uniform grid from 0 to output size, the user specifies it instead.

**Comment by soumith** [comment]
yea, this feature request is very very different from "interpolation for images".

**Comment by zhangguanheng66** [comment]
After taking a look at the feature request, it seems that the unstructured interpolation function is requested. I will first work on the unstructured interpolation on 1d.

**Comment by peteflorence** [comment]
Hi folks, the code that I wrote up before does do bilinear interp on a
non-regular grid, in case useful... (and I am actively using it in a
project of mine for C-dimensional images)
https://gist.github.com/peteflorence/a1da2c759ca1ac2b74af9a83f69ce20e

https://gist.github.com/peteflorence/a1da2c759ca1ac2b74af9a83f69ce20e

**Comment by jiangyangzhou** [comment]
So if unstructured 2-D interpolation can be implement in the future?  scipy.interpolate.griddata performs unstructured n-D interpolation,  and I need this function in my project.  It would be better if Pytorch can add this function.

**Comment by zhangguanheng66** [comment]
@jiangyangzhou I have a 1-D interpolation function on CPU (see the PR above).

**Comment by jiangyangzhou** [comment]
However, I need a 2-D interpolation, I have some scatter points on image, and I want to interpolate to whole image. I found scipy.interpolate.griddata is a solution, but I need to compute gradient. So I am looking for GPU solution.  It seems I should triangulate the input data and implement interpolation.

**Comment by jph00** [comment]
This is very simply to implement in plain pytorch if we have a `searchsorted` function like numpy. Here's a memory-inefficient implementation in plain pytorch:

```python
def interp_1d(x, xp, fp):
    "Same as `np.interp`"
    slopes = (fp[1:]-fp[:-1])/(xp[1:]-xp[:-1])
    incx = fp[:-1] - (slopes*xp[:-1])
    locs = (x[:,None]>=xp[None,:]).long().sum(1)-1
    return slopes[locs]*x + incx[locs]
```

The line that calculates `locs` would be much faster by using `searchsorted`. Here's some related links:

- Open issue for interp: https://github.com/pytorch/pytorch/issues/1552
- Open PR for interp (only CPU): https://github.com/pytorch/pytorch/pull/20714
- Another feature request that could leverage searchsorted: https://github.com/pytorch/pytorch/issues/7284
- 3rd party implementation of searchsorted: https://github.com/aliutkus/torchsearchsorted

cc @soumith

**Comment by aliutkus** [comment]
well, as I mentioned above, I implemented a 1D linear implementation for pytorch already:
[https://github.com/aliutkus/torchinterp1d](https://github.com/aliutkus/torchinterp1d)

**Comment by jph00** [comment]
Yes this is already included in the links in the above comment.

But it would be very helpful to have this functionality as part of PyTorch, rather than requiring an external dependency, that needs to be compiled and packaged for each target environment.

**Comment by aliutkus** [comment]
please note it's not, I linked here to an implementation of unstructured 1d linear interpolation, not to the searchsorted function (which is indeed a dependency)

**Comment by jph00** [comment]
Ah yes sorry for the mix up!

**Comment by patrick-kidger** [comment]
If it helps anyone who got here by Googling (like I did) - I've ended up writing my own approach for performing natural cubic spline interpolation:

[torchcubicspline](https://github.com/patrick-kidger/torchcubicspline)

It includes support for:
- Batching
- Missing values
- GPU support and backprop via PyTorch
- The ability to directly evaluate the first derivative of the spline.

**Comment by sbarratt** [comment]
I implemented this using torch functions, so it's automatically  GPU compatible & fully differentiable. I could probably port it to C++ and make it a bit faster. https://github.com/sbarratt/torch_interpolations

For interpolating 4.5 million points on a 300 x 300 grid, it takes:
```
PyTorch took 222.531 +\- 7.972 ms
PyTorch Cuda took 12.502 +\- 0.493 ms
Scipy took 421.471 +\- 4.278 ms
```
(on a 1080 TI and i7-8700K.)

**Comment by lc82111** [comment]
Below link may be helpful in the irregular grid interpolation.
https://www.semicolonworld.com/question/59331/speedup-scipy-griddata-for-multiple-interpolations-between-two-irregular-grids

**Comment by mruberry** [comment]
Thank you everyone who participated in this issue. Since it's an older issue and covers many requests, I've filed more modern issues tracking each request. It's helpful when issues map to a specific request.

First, the good news: PyTorch now has [torch.searchsorted](https://pytorch.org/docs/master/generated/torch.searchsorted.html?highlight=searchsorted#torch.searchsorted)!

Issues to track the other requests made in this thread:

- https://github.com/pytorch/pytorch/issues/50334 for np.interp
- https://github.com/pytorch/pytorch/issues/50335 for scipy.interpolate.interp1d
- https://github.com/pytorch/pytorch/issues/50336 for scipy.ndimage.zoom
- https://github.com/pytorch/pytorch/issues/50337 for scipy.interpolate.RectBivariateSpline
- https://github.com/pytorch/pytorch/issues/50338 for scipy.interpolate.RegularGridInterpolator
- https://github.com/pytorch/pytorch/issues/50339 for scipy.interpolate.griddata
- https://github.com/pytorch/pytorch/issues/50340 for scipy.interpolate.InterpolatedUnivariateSpline

I believe that captures all the requests? Please let me know if I'm mistaken, and I encourage everyone to continue filing issues requesting NumPy or SciPy (or MatLab or...) functions that would be helpful when writing PyTorch programs. 

I also created a tracking issue for interpolation in PyTorch here https://github.com/pytorch/pytorch/issues/50341, so that we can continue to have one place to review and discuss these related issues and ideas.

Closing this issue in favor of the new specific issues and the tracking issue.

**Comment by ZichaoLong** [comment]
Here is a implementation of N-D Lagrange interpolation. 
Just like scipy.interpolate.interpn, it supports irregular input grid(and requires values on regular grid). 
It support arbitrary interpolation order.

https://github.com/ZichaoLong/aTEAM/blob/master/nn/modules/Interpolation.py
https://github.com/ZichaoLong/aTEAM/blob/master/nn/functional/interpolation.py

**Comment by TobiasJacob** [comment]
In case anybody is looking for this in a functional form, here is a 4-liner:

```python
import torch
xp = torch.arange(0, 1, 0.1) # X values
fp = torch.arange(0, 10, 1.0) # Y values

x = torch.Tensor([[0.1, 0.05, 0.3], [1.5, -2, 0.5]]) # X values you need to interpolate

def interpolate(xp, fp, x):
    slopes = (fp[1:]-fp[:-1])/(xp[1:]-xp[:-1])
    locs = torch.searchsorted(xp, x)
    locs = locs.clip(1, len(xp) - 1) - 1
    return slopes[locs] * (x - xp[locs]) + xp[locs]

interpolate(xp, fp, x)
```

**Comment by perrette** [comment]
@TobiasJacob thanks for sharing, however, it seems it does not work great in edge cases (like on the points):

`interpolate(torch.arange(5), torch.arange(5)**2, 3)` yields 7... (instead of expected 9, as `np.interp` correctly yields...).

**Comment by perrette** [comment]
The above function is definitely wrong. 
Here a two liners that works, adapted from https://github.com/google/jax/issues/3860#issuecomment-663866144:

    def interpolate(x, xp, fp):
        i = torch.clip(torch.searchsorted(xp, x, right=True), 1, len(xp) - 1)
        return (fp[i - 1] *  (xp[i] - x) + fp[i] * (x - xp[i - 1])) / (xp[i] - xp[i - 1])

**Comment by cjang105** [comment]
> Well actually, you can also use np.interp() to process torch type data. But the type of return value is numpy. Then you change it to torch type~

It will lose gradients and the model will be not differentiable anymore, which is most likely not desired

**Comment by eigenvivek** [comment]
might be useful for some cases: `pytorch_geometric` has a `knn_interpolation` function https://pytorch-geometric.readthedocs.io/en/stable/generated/torch_geometric.nn.unpool.knn_interpolate.html#torch_geometric.nn.unpool.knn_interpolate
